tests/integration/relations/pgbouncer_provider/application-charm/src/charm.py

```diff
@@ -0,0 +1,161 @@
+        )
+
+        # Events related to the second database that is requested
+        # (these events are defined in the database requires charm library).
+        database_name = f'{self.app.name.replace("-", "_")}_second_database'
+        self.second_database = DatabaseRequires(
+            self, "second-database", database_name, EXTRA_USER_ROLES
+        )
+        self.framework.observe(
+            self.second_database.on.database_created, self._on_second_database_created
+        )
+        self.framework.observe(
+            self.second_database.on.endpoints_changed, self._on_second_database_endpoints_changed
+        )
+
+        self.framework.observe(self.on.run_sql_action, self._on_run_sql_action)
+
+    def _on_start(self, _) -> None:
+        """Only sets an Active status."""
+        self.unit.status = ActiveStatus()
+
+    # First database events observers.
+    def _on_first_database_created(self, event: DatabaseCreatedEvent) -> None:
+        """Event triggered when a database was created for this application."""
+        # Retrieve the credentials using the charm library.
+        logger.info(f"first database credentials: {event.username} {event.password}")
+        self.unit.status = ActiveStatus("received database credentials of the first database")
+
+    def _on_first_database_endpoints_changed(self, event: DatabaseEndpointsChangedEvent) -> None:
+        """Event triggered when the read/write endpoints of the database change."""
+        logger.info(f"first database endpoints have been changed to: {event.endpoints}")
+
+    # Second database events observers.
+    def _on_second_database_created(self, event: DatabaseCreatedEvent) -> None:
+        """Event triggered when a database was created for this application."""
+        # Retrieve the credentials using the charm library.
+        logger.info(f"second database credentials: {event.username} {event.password}")
+        self.unit.status = ActiveStatus("received database credentials of the second database")
+
+    def _on_second_database_endpoints_changed(self, event: DatabaseEndpointsChangedEvent) -> None:
+        """Event triggered when the read/write endpoints of the database change."""
+        logger.info(f"second database endpoints have been changed to: {event.endpoints}")
+
+    def _on_run_sql_action(self, event: ActionEvent):
+        """An action that allows us to run SQL queries from this charm."""
+        logger.info(event.params)
+
+        relation_id = event.params["relation-id"]
+        relation_name = event.params["relation-name"]
+        if relation_name == self.first_database.relation_name:
+            relation = self.first_database
+        elif relation_name == self.second_database.relation_name:
+            relation = self.second_database
+        elif relation_name == self.database_clusters.relation_name:
+            relation = self.database_clusters
```

Comment:
Does `self.database_clusters` exist? I couldn't find it in this charm code.

Reply:
nope, forgot to remove this. Good catch!